File: javascript/src/factory-method/factory-method.test.js
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { ConcreteCreatorA } from "./concrete-creator-a.js";
import { ConcreteCreatorB } from "./concrete-creator-b.js";
import { Creator } from "./creator.js";

describe("FactoryMethod", () => {
  it("each concrete creator builds its own product", () => {
    assert.equal(new ConcreteCreatorA().anOperation(), "Built ConcreteProductA");
    assert.equal(new ConcreteCreatorB().anOperation(), "Built ConcreteProductB");
  });

  it("template operation uses whatever the subclass returns", () => {
    class CustomCreator extends Creator {
      factoryMethod() {
        return { name: () => "Custom" };
      }
    }
    assert.equal(new CustomCreator().anOperation(), "Built Custom");
  });

  it("a creator without factoryMethod() fails at the first call", () => {
    assert.throws(() => new Creator().anOperation(), /must implement factoryMethod/);
  });
});

File: java/src/test/java/com/penapereira/patterns/factorymethod/FactoryMethodTest.java
package com.penapereira.patterns.factorymethod;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

class FactoryMethodTest {

    @Test
    void eachConcreteCreatorBuildsItsOwnProduct() {
        assertEquals("Built ConcreteProductA", new ConcreteCreatorA().anOperation());
        assertEquals("Built ConcreteProductB", new ConcreteCreatorB().anOperation());
    }

    @Test
    void templateOperationUsesWhateverTheSubclassReturns() {
        Creator custom = new Creator() {
            @Override
            protected Product factoryMethod() {
                return () -> "Custom";
            }
        };
        assertEquals("Built Custom", custom.anOperation());
    }
}
